lib/std/hash_map.zig

```diff
@@ -1503,6 +1509,94 @@ pub fn HashMapUnmanaged(
             return result;
         }
 
+        /// Rehash the map, in-place.
+        ///
+        /// Over time, due to the current tombstone-based implementation, a
+        /// HashMap could become fragmented due to the buildup of tombstone
+        /// entries that causes a performance degradation due to excessive
+        /// probing. The kind of pattern that might cause this is a long-lived
+        /// HashMap with repeated inserts and deletes.
+        ///
+        /// All existing key/value pointers in the HashMap are maintained at
+        /// their new rehashed location.
```

Comment:
Sorry, I'm still not quite clear on this, does it invalidate existing pointers or not?

> All existing key/value pointers in the HashMap are maintained

Sounds like yes

> new rehashed location

Sounds like no

Reply:
I maybe don't know enough about Zig, we do this:

```
            var keys_ptr = self.keys();
            var values_ptr = self.values();
```

and later, to store the pointer to its new rehashed location:

```
                    assert(metadata[idx].isFree());
                    metadata[idx].fill(fingerprint);
                    keys_ptr[idx] = keys_ptr[curr];
                    values_ptr[idx] = values_ptr[curr];
```

the pointers are moved in the array but their value is preserved.

Reply:
And also using std.mem.swap

```
                    if (metadata[idx].isUsed()) {
                        std.mem.swap(K, &keys_ptr[curr], &keys_ptr[idx]);
                        std.mem.swap(V, &values_ptr[curr], &values_ptr[idx]);
                    } else {
                        metadata[idx].used = 1;
                        keys_ptr[idx] = keys_ptr[curr];
                        values_ptr[idx] = values_ptr[curr];
                    }
```

Reply:
I'll ask the question with code instead:

```zig
const a_key_ptr = &map.keys()[i];
const a_value_ptr = &map.values()[i];
map.rehash();
foo(a_key_ptr.*);
bar(a_value_ptr.*);
```

does this invoke illegal behavior?

the answer to this question is not communicated clearly in the doc comments.

Reply:
I'll give you a hint: it definitely invalidates existing pointers

Reply:
Yes, if you have ``a_key_ptr`` as a pointer into the keys array, the value that points to will change after rehashing.

And that needs to be reflected in the documentation.

Reply:
> Yes, if you have `a_key_ptr` as a pointer into the keys array, the value that points to will (could) change after rehashing.

Incorrect. The value pointed to will remain the same. The pointer is invalidated.

Reply:
Is my ``rehash()`` approach flawed, or do I just need to document that this occurs?

Reply:
It's expected that a rehash will invalidate live pointers, so documenting it should be enough, as other functions modifying the hashmap do.

Reply:
Hi @Sahnvour and @andrewrk, thank you. I have updated the docs.

I apologize my understanding of Zig isn't deep enough yet to recognize the issue of invalidating live pointers into the map.